v2/pkg/genruntime/extensions/prereconciliation_checker.go

```diff
@@ -0,0 +1,167 @@
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+)
+
+// PreReconciliationChecker is implemented by resources that want to do extra checks before proceeding with
+// a full ARM reconcile.
+type PreReconciliationChecker interface {
+	// PreReconcileCheck does a pre-reconcile check to see if the resource is in a state that can be reconciled.
+	// ARM resources should implement this to avoid reconciliation attempts that cannot possibly succeed.
+	// Returns ProceedWithReconcile if the reconciliation should go ahead.
+	// Returns BlockReconcile and a human-readable reason if the reconciliation should be skipped.
+	// ctx is the current operation context.
+	// obj is the resource about to be reconciled. The resource's State will be freshly updated.
+	// kubeClient allows access to the cluster for any required queries.
+	// armClient allows access to ARM for any required queries.
+	// log is the logger for the current operation.
+	// next is the next (nested) implementation to call.
+	PreReconcileCheck(
+		ctx context.Context,
+		obj genruntime.MetaObject,
+		owner genruntime.MetaObject,
+		kubeClient kubeclient.Client,
+		armClient *genericarmclient.GenericClient,
+		log logr.Logger,
+		next PreReconcileCheckFunc,
+	) (PreReconcileCheckResult, error)
+}
+
+type PreReconcileCheckFunc func(
+	ctx context.Context,
+	obj genruntime.MetaObject,
+	owner genruntime.MetaObject,
+	kubeClient kubeclient.Client,
+	armClient *genericarmclient.GenericClient,
+	log logr.Logger,
+) (PreReconcileCheckResult, error)
+
+type PreReconcileCheckResult struct {
+	action   preReconcileCheckResultType
+	severity conditions.ConditionSeverity
+	reason   conditions.Reason
+	message  string
+}
+
+// ProceedWithReconcile indicates that a resource is ready for reconciliation by returning a PreReconcileCheckResult
+// with action `Proceed`.
+func ProceedWithReconcile() PreReconcileCheckResult {
+	return PreReconcileCheckResult{
+		action: preReconcileCheckResultTypeProceed,
+	}
+}
+
+// BlockReconcile indicates reconciliation of a resource is currently blocked by returning a PreReconcileCheckResult
+// with action `Block`.
+// reason is an explanatory reason to show to the user via a warning condition on the resource.
+func BlockReconcile(reason string) PreReconcileCheckResult {
+	return PreReconcileCheckResult{
+		action:   preReconcileCheckResultTypeBlock,
+		severity: conditions.ConditionSeverityWarning,
+		reason:   conditions.ReasonReconcileBlocked,
+		message:  reason,
+	}
+}
+
+// PostponeReconcile indicates reconciliation of a resource is not currently required by returning a
+// PreReconcileCheckResult with action `Postpone`.
+// reason is an explanatory reason to show to the user via an info condition on the resource.
+func PostponeReconcile(reason string) PreReconcileCheckResult {
+	return PreReconcileCheckResult{
+		action:   preReconcileCheckResultTypePostpone,
+		severity: conditions.ConditionSeverityInfo,
```

Comment:
Ah, for `Postpone`, if we just want to allow the standard "eventually we will re-check" interval, then we shouldn't return an error here.

The success case (== no reconcile needed at this time, try again after the standard `azure_sync_period` is `ctrl.Result{}, nil`). If you return an err, you'll end up getting reconciled again after only a small interval (3m is our max backoff I think), issuing another GET to Azure.
The block case should be `ctrl.Result{}, err` (which you already have).

Reply:
So how do we show the reason for the postponement on the condition for the resource? I thought returning an error was required here.

Reply:
As discussed offline, I agree that "success is success", so we don't need to show a reason for postponement.